lib/record_store/provider.rb

```diff
@@ -3,14 +3,19 @@
 module RecordStore
   class Provider
     class << self
-      def provider_for(zone_name)
-        dns = Resolv::DNS.new(nameserver: ['8.8.8.8', '8.8.4.4'])
-
-        begin
-          ns_server = dns.getresource(zone_name, Resolv::DNS::Resource::IN::SOA).mname.to_s
-        rescue Resolv::ResolvError
-          abort("Domain doesn't exist")
-        end
+      def provider_for(object)
+        ns_server =
+          case object
+          when Record::NS
+            object.nsdname.chomp('.')
+          else
+            begin
+              master_nameserver_for(object)
+            rescue Resolv::ResolvError
+              $stderr.puts "Domain doesn't exist (#{object})"
+              return
+            end
+          end
```

Comment:
Why a case here instead of simple `if`?

Reply:
to easily match the types... and I'd originally considered `when String` for the current else case

I can definitely switch to if `object.is_a?(Record::NS)` instead 🤷